### PR TITLE
Caching blue rabbit

### PR DESCRIPTION
**Pull request by andrey-prokopyev**
Merged improved caching from silver-lizard.
#100

### CODE REVIEW

File: Sources/Contour/Caching/CacheConfigProvider.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+namespace Contour.Caching
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Configuration;
+    using System.Linq;
+    using System.Xml.Linq;
+
+    internal class CacheConfigProvider
```
Comment (mzabolotko) [inline Sources/Contour/Caching/CacheConfigProvider.cs]: I think this class should be placed to the Contour.Configurator namespace.

File: Sources/Contour/Caching/ThreadSafeCache.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        public void Set(IMessage key, object value, TimeSpan ttl)
+        {
+            this.locker.EnterWriteLock();
+            try
+            {
+                this.innerCache.Set(key, value, ttl);
+            }
+            finally
+            {
+                this.locker.EnterWriteLock();
```
Comment (mzabolotko) [inline Sources/Contour/Caching/ThreadSafeCache.cs]: spelling mistake?

File: Sources/Contour/Caching/ThreadSafeCache.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
+namespace Contour.Caching
+{
+    using System;
+    using System.Threading;
+
+    public class ThreadSafeCache : ICache
```
Comment (mzabolotko) [inline Sources/Contour/Caching/ThreadSafeCache.cs]: I think it is out of scope of Contour.

File: Sources/Contour/Caching/CacheConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+﻿namespace Contour.Caching
+{
+    using System;
+
+    public class CacheConfiguration
```
Comment (mzabolotko) [inline Sources/Contour/Caching/CacheConfiguration.cs:5]: Should be internal.

File: Sources/Contour/Caching/CacheMessageExchangeFilter.cs [hunk 5]
```diff
@@ -1,82 +0,0 @@
...
-
-using Contour.Filters;
-using Contour.Helpers;
-
-namespace Contour.Caching
-{
-    /// <summary>
-    /// Фильтр, который кеширует ответные сообщения.
-    /// </summary>
-    public class CacheMessageExchangeFilter : IMessageExchangeFilter
```
Comment (mzabolotko) [inline Sources/Contour/Caching/CacheMessageExchangeFilter.cs]: Breaking backward compatibility. You should avoid it.

File: Sources/Contour/Caching/ICacheProvider.cs [hunk 6]
```diff
@@ -1,82 +0,0 @@
-﻿namespace Contour.Caching
-{
-    using System;
-
-    using Contour.Helpers;
-
-    /// <summary>
-    /// The CacheProvider interface.
-    /// </summary>
-    public interface ICacheProvider
```
Comment (mzabolotko) [inline Sources/Contour/Caching/ICacheProvider.cs]: Breaking backward compatibility. You should avoid it.

File: Sources/Contour/Caching/MemoryCacheProvider.cs [hunk 7]
```diff
@@ -1,153 +0,0 @@
...
-    using System;
-    using System.Runtime.Caching;
-
-    using Contour.Helpers;
-    using Contour.Helpers.CodeContracts;
-
-    /// <summary>
-    /// The memory cache provider.
-    /// </summary>
-    public class MemoryCacheProvider : ICacheProvider
```
Comment (mzabolotko) [inline Sources/Contour/Caching/MemoryCacheProvider.cs]: Breaking backward compatilibity. You should avoid it.

File: Sources/Contour/Configuration/IBusConfigurator.cs [hunk 8]
```diff
@@ -16,11 +16,6 @@ namespace Contour.Configuration
     /// </summary>
     public interface IBusConfigurator
     {
-        /// <summary>
-        /// The enable caching.
-        /// </summary>
-        void EnableCaching();
```
Comment (mzabolotko) [inline Sources/Contour/Configuration/IBusConfigurator.cs]: Breaking backward compatibility. You should avoid it.

File: Sources/Contour/Helpers/Hasher.cs [hunk 9]
```diff
@@ -1,39 +0,0 @@
-﻿using System;
-
-using Newtonsoft.Json;
-
-namespace Contour.Helpers
-{
-    /// <summary>
-    /// Рассчитывает значение хеш-функции для содержимого.
-    /// </summary>
-    public class Hasher
```
Comment (mzabolotko) [inline Sources/Contour/Helpers/Hasher.cs]: Breaking backward compatibility. You should avoid it.

**Review by mzabolotko — CHANGES_REQUESTED** [review]
I think you should avoid backward compatibility breaking. Please, fix spelling mistake. Please remove  unnecessary code.

File: Sources/Contour/Caching/CacheConfigProvider.cs (hunk 1 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Caching/CacheConfigProvider.cs]: moved

File: Sources/Contour/Caching/ThreadSafeCache.cs (hunk 2 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Caching/ThreadSafeCache.cs]: deleted this class

File: Sources/Contour/Caching/ThreadSafeCache.cs (hunk 3 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Caching/ThreadSafeCache.cs]: I think so, Deleted this class

File: Sources/Contour/Caching/CacheConfiguration.cs (hunk 4 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Caching/CacheConfiguration.cs:5]: CacheConfiguration is a constructor dependency for public CachingConsumerOf class, so it cannot be internal.

File: Sources/Contour/Caching/CacheMessageExchangeFilter.cs (hunk 5 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Caching/CacheMessageExchangeFilter.cs]: Reverted deletion and marked as obsolete.

File: Sources/Contour/Caching/ICacheProvider.cs (hunk 6 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Caching/ICacheProvider.cs]: Reverted deletion and marked as obsolete.

File: Sources/Contour/Caching/MemoryCacheProvider.cs (hunk 7 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Caching/MemoryCacheProvider.cs]: Reverted deletion and marked as obsolete.

File: Sources/Contour/Configuration/IBusConfigurator.cs (hunk 8 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Configuration/IBusConfigurator.cs]: Reverted deletion and marked as obsolete.

File: Sources/Contour/Helpers/Hasher.cs (hunk 9 above)
Comment (andrey-prokopyev) [reply Sources/Contour/Helpers/Hasher.cs]: Reverted deletion and marked as obsolete.

**Pull request merged**